tee-worker/litentry/core/data-providers/src/nodereal_jsonrpc.rs

```diff
@@ -0,0 +1,343 @@
+pub struct ReqPath {
+	path: String,
+}
+
+impl ReqPath {
+	pub fn new(api_key: &str) -> Self {
+		Self { path: "v1/".to_string() + api_key }
+	}
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct ReqBody<P>
+where
+	P: Serialize,
+{
+	pub id: i64,
+	pub jsonrpc: String,
+	pub method: String,
+	pub params: P,
+}
+
+impl<P> ReqBody<P>
+where
+	P: Serialize,
+{
+	fn new(jsonrpc: &str, method: &str, params: P) -> Self {
+		ReqBody { id: 1, jsonrpc: jsonrpc.to_string(), method: method.to_string(), params }
+	}
+}
+
+impl<P> RestPath<ReqPath> for ReqBody<P>
+where
+	P: Serialize,
+{
+	fn get_path(req: ReqPath) -> core::result::Result<String, HttpError> {
+		Ok(req.path)
+	}
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct RespBody {
+	pub id: i64,
+	pub jsonrpc: String,
+	pub result: serde_json::Value,
+}
+
+pub struct NoderealJsonrpcClient {
+	api_key: String,
+	api_retry_delay: u64,
+	api_retry_times: u16,
+	client: RestClient<HttpClient<DefaultSend>>,
+}
+
+impl NoderealJsonrpcClient {
+	pub fn new(chain: NoderealChain) -> Self {
+		let api_key = GLOBAL_DATA_PROVIDER_CONFIG.write().unwrap().nodereal_api_key.clone();
+		let api_retry_delay = GLOBAL_DATA_PROVIDER_CONFIG.write().unwrap().nodereal_api_retry_delay;
+		let api_retry_times = GLOBAL_DATA_PROVIDER_CONFIG.write().unwrap().nodereal_api_retry_times;
+		let api_url = GLOBAL_DATA_PROVIDER_CONFIG
+			.write()
+			.unwrap()
+			.nodereal_api_chain_network_url
+			.clone();
+		let base_url =
+			api_url.replace("{chain}", chain.to_string()).replace("{network}", "mainnet");
+
+		let mut headers = Headers::new();
+		headers.insert(CONNECTION.as_str(), "close");
+		let client = build_client(base_url.as_str(), headers);
+
+		NoderealJsonrpcClient { api_key, api_retry_delay, api_retry_times, client }
+	}
+
+	// https://docs.nodereal.io/docs/cups-rate-limit
+	// add retry functionality to handle situations where requests may surpass predefined constraints.
+	fn retry<A>(&mut self, action: A) -> Result<RespBody, Error>
+	where
+		A: Fn(&mut NoderealJsonrpcClient) -> Result<RespBody, HttpError>,
+	{
+		let mut retries = 0;
+		// retry delay 1 second
+		let base_delay = time::Duration::from_millis(self.api_retry_delay);
+		// maximum 5 retry times
+		let maximum_retries = self.api_retry_times;
+
+		loop {
+			if retries > 0 {
+				debug!("Fail to call nodereal enhanced api, begin retry: {}", retries);
+			}
+
+			if retries > maximum_retries {
+				return Err(Error::RequestError(format!(
+					"Fail to call call nodereal enhanced api within {} retries",
+					maximum_retries
+				)))
+			}
+
+			match action(self) {
+				Ok(response) => return Ok(response),
+				Err(err) => {
+					let req_err: Error =
+						Error::RequestError(format!("Nodereal enhanced api error: {}", err));
+					match err {
+						HttpError::HttpError(code, _) =>
+							if code == 429 {
+								// Too Many Requests
+								// exponential back off
+								thread::sleep(base_delay * 2u32.pow(retries as u32));
```

Comment:
This will potentially block the whole VC issuance thread for 32s - this can easily be an attacking interface without threadpool and VC-Isolation. 

Even if that's not intentional - people could spam the requests if they don't get requests for half a minute which would make it even worse 😂 

How likely will this 429 happen? Can we only retry for 1-2 times and then error out (i.e. ask the user to retry it later)

Reply:
Yes, 1-2 times would be enough. It's better to ask user to retry it from IDHub.

Reply:
This will make this more complicated but what about not blocking thread while waiting for retry ? 
Can this request go back to the queue ?